jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp

```diff
@@ -0,0 +1,204 @@
+    return false;
+
+  const typet &decl_type = to_code_decl(to_code(expr)).symbol().type();
+  return decl_type.id() == ID_struct_tag &&
+         to_struct_tag_type(decl_type).get_identifier() == id;
+}
+
+static bool is_decl_of_type(const exprt &expr, const typet &type)
+{
+  return expr.id() == ID_code && to_code(expr).get_statement() == ID_decl &&
+         to_code_decl(to_code(expr)).symbol().type() == type;
+}
+
+static bool
+contains(const exprt &expr, std::function<bool(const exprt &)> predicate)
+{
+  if(predicate(expr))
+    return true;
+  for(const auto &op : expr.operands())
+  {
+    if(contains(op, predicate))
+      return true;
+  }
+  return false;
+}
+
+static bool
+contains(const typet &type, std::function<bool(const typet &)> predicate)
+{
+  if(predicate(type))
+    return true;
+
+  if(type.has_subtypes())
+  {
+    for(const auto &subtype : to_type_with_subtypes(type).subtypes())
+    {
+      if(contains(subtype, predicate))
+        return true;
+    }
+  }
+
+  return false;
+}
+
+static bool contains_decl_of_type(const exprt &expr, const typet &type)
+{
+  return contains(expr, [&type](const exprt &subexpr) {
+    return is_decl_of_type(subexpr, type);
+  });
+}
+
+static bool contains_decl_with_id(const exprt &expr, const irep_idt &id)
+{
+  return contains(
+    expr, [&id](const exprt &subexpr) { return is_decl_with_id(subexpr, id); });
+}
+
+static bool uses_raw_struct_types(const typet &type)
+{
+  return contains(
+    type, [](const typet &subtype) { return subtype.id() == ID_struct; });
+}
+
+static bool uses_raw_struct_types(const exprt &expr)
+{
+  return contains(expr, [](const exprt &subexpr) {
+    return uses_raw_struct_types(subexpr.type());
+  });
+}
+
+code_blockt
+initialise_nondet_object_of_type(const typet &type, symbol_tablet &symbol_table)
+{
+  code_blockt created_code;
+  java_object_factory_parameterst parameters;
+  select_pointer_typet pointer_selector;
+
+  object_factory(
+    type,
+    "root",
+    created_code,
+    symbol_table,
+    parameters,
+    lifetimet::AUTOMATIC_LOCAL,
+    source_locationt(),
+    pointer_selector);
+
+  return created_code;
+}
+
+SCENARIO(
+  "java_object_factory_tag_types",
+  "[core][java_bytecode][java_object_factory]")
+{
+  GIVEN("Some classes with fields")
+  {
+    register_language(new_java_bytecode_language);
+    symbol_tablet symbol_table =
+      load_java_class("Root", "./java_bytecode/java_object_factory");
+
+    WHEN("Creating a nondet 'A' object")
+    {
+      struct_tag_typet A_type("java::A");
+      struct_tag_typet B_type("java::B");
+      struct_tag_typet C_type("java::C");
+
+      const auto a_pointer = java_reference_type(A_type);
+      code_blockt created_code =
+        initialise_nondet_object_of_type(a_pointer, symbol_table);
+
+      THEN("An A, a B and a C object should be allocated")
+      {
+        REQUIRE(contains_decl_of_type(created_code, A_type));
```

Comment:
:bulb: You might like to use `require_goto_statements::require_declaration_of_name` either here, or inline, or expand it with your functionality.